pkg/webhook/admission/defaulter_custom.go

```diff
@@ -21,29 +21,51 @@ import (
 	"encoding/json"
 	"errors"
 	"net/http"
+	"slices"
 
+	"gomodules.xyz/jsonpatch/v2"
 	admissionv1 "k8s.io/api/admission/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
 )
 
 // CustomDefaulter defines functions for setting defaults on resources.
 type CustomDefaulter interface {
 	Default(ctx context.Context, obj runtime.Object) error
 }
 
+type defaulterOptions struct {
+	removeUnknownFields bool
+}
+
+// DefaulterOption defines the type of a CustomDefaulter's option
+type DefaulterOption func(*defaulterOptions)
+
+// DefaulterRemoveUnknownFields makes the defaulter prune fields that are in the json object retrieved by the
+// webhook but not in the local go type. This happens for example when the CRD in the apiserver has fields that
+// our go type doesn't know about, because it's outdated.
+func DefaulterRemoveUnknownFields(o *defaulterOptions) {
```

Comment:
Not sure if the name of this option is correct. It will remove more than just unknown fields, right?

It will remove all fields that would be dropped by json.Unmarshal + json.Marshal.

If I see correctly this would also affect for example fields with omitempty. An example:

Type
```
   // +optional
   MinReadySeconds int32 `json:"minReadySeconds,omitempty"`
```

Incoming object:
```
   minReadySeconds: 0
```

After json.Unmarshal + json.Marshal this field would also be removed.

So this option removes more than just unknown fields, right?

(I"m aware that it just enables the previous behavior, but I'm wondering if the name is misleading)

Reply:
How about something like:

```go
// DefaulterRemoveUnknownOrOmitableFields makes the defaulter prune fields that are in the json object retrieved by the
// webhook but not in the local go type json representation. This happens for example when the CRD in the apiserver has
// fields that our go type doesn't know about, because it's outdated, the field has a zero value and is `omitempty`.
func DefaulterRemoveUnknownOrOmitableFields(o *defaulterOptions) {

```

Reply:
Updated

Reply:
Thx!